crates/component-macro/src/lib.rs

```diff
@@ -0,0 +1,382 @@
+        syn::Fields::Named(fields) => {
+            let mut types = TokenStream::new();
+            let mut field_initializations = TokenStream::new();
+
+            for (index, syn::Field { ty, ident, .. }) in fields.named.iter().enumerate() {
+                types.extend(quote!(#ty,));
+
+                let index = syn::Index::from(index);
+
+                field_initializations.extend(quote!(#ident: result.#index,));
+            }
+
+            delegate_record_lift(
+                input,
+                quote!((#types)),
+                quote!(Self { #field_initializations }),
+            )
+        }
+        syn::Fields::Unnamed(_) | syn::Fields::Unit => {
+            Err(Error::new(name.span(), INCOMPATIBLE_TYPE_MESSAGE))
+        }
+    }
+}
+
+fn delegate_record_lift(
+    input: DeriveInput,
+    target: TokenStream,
+    lift: TokenStream,
+) -> Result<TokenStream> {
+    let internal = quote!(wasmtime::component::__internal);
+    let name = input.ident;
+    let generics = add_trait_bounds(input.generics);
+    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
+
+    let expanded = quote! {
+        unsafe impl #impl_generics wasmtime::component::Lift for #name #ty_generics #where_clause {
+            #[inline]
+            fn lift(
+                store: &#internal::StoreOpaque,
+                options: &#internal::Options,
+                src: &Self::Lower,
+            ) -> #internal::anyhow::Result<Self> {
+                let result = <#target as wasmtime::component::Lift>::lift(store, options, src)?;
+                Ok(#lift)
+            }
+
+            #[inline]
+            fn load(mem: &#internal::Memory, bytes: &[u8]) -> #internal::anyhow::Result<Self> {
+                let result = <#target as wasmtime::component::Lift>::load(mem, bytes)?;
+                Ok(#lift)
+            }
+        }
+    };
+
+    Ok(expanded)
+}
+
+#[proc_macro_derive(Lower, attributes(component))]
+pub fn lower(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    expand_lower(parse_macro_input!(input as DeriveInput))
+        .unwrap_or_else(Error::into_compile_error)
+        .into()
+}
+
+fn expand_lower(input: DeriveInput) -> Result<TokenStream> {
+    match find_style(&input)? {
+        Style::Record => expand_record_lower(input),
+
+        style => Err(Error::new_spanned(
+            input,
+            format!("todo: expand_lower {style:?}"),
+        )),
+    }
+}
+
+fn expand_record_lower(input: DeriveInput) -> Result<TokenStream> {
+    let name = &input.ident;
+
+    let body = if let Data::Struct(body) = &input.data {
+        body
+    } else {
+        return Err(Error::new(name.span(), INCOMPATIBLE_TYPE_MESSAGE));
+    };
+
+    match &body.fields {
+        syn::Fields::Named(fields) => {
+            let values = fields
+                .named
+                .iter()
+                .map(|syn::Field { ident, .. }| quote!(self.#ident.clone(),))
```

Comment:
Instead of cloning could this pass `&T` since `&T` is `Lower` if `T` is `Lower`?